lib/pbench/agent/tool_meister_client.py

```diff
@@ -87,149 +83,126 @@ def __init__(
+        assert publisher_prefix is not None, "You must specify a publisher prefix"
+        self.publisher_name = f"{publisher_prefix}-pbench-client"
 
         if logger is None:
             self.logger = logging.getLogger("tool-meister-client")
         else:
             self.logger = logger
 
     def __enter__(self) -> "Client":
-        """On context entry, setup the connection with the Redis server."""
-        if self.redis_server is None:
-            self.logger.debug("constructing Redis() object")
-            try:
-                self.redis_server = redis.Redis(
-                    host=self.redis_host, port=self.redis_port, db=0
-                )
-            except Exception as e:
-                self.logger.error(
-                    "Unable to connect to redis server, %s:%s: %s",
-                    self.redis_host,
-                    self.redis_port,
-                    e,
-                )
-                raise
-            else:
-                self.logger.debug("constructed Redis() object")
-
-        if self.to_client_chan is None:
-            self.to_client_chan = RedisChannelSubscriber(
-                self.redis_server, self.to_client_channel
+        """On context entry, setup the connection with the Redis server using
+        a SignalExporter instance."""
+        if self.redis_server:
+            self.logger.debug(
+                "constructing SignalExporter() object using existing Redis"
+                " connection, name: %s",
+                self.publisher_name,
+            )
+            sig_pub = state_signals.SignalExporter(
+                self.publisher_name, existing_redis_conn=self.redis_server
+            )
+        else:
+            self.logger.debug(
+                "constructing SignalExporter() object using host %s:%s," " name: %s",
+                self.redis_host,
+                self.redis_port,
+                self.publisher_name,
             )
+            sig_pub = state_signals.SignalExporter(
+                self.publisher_name,
+                redis_host=self.redis_host,
+                redis_port=self.redis_port,
+            )
+        sig_pub.initialize_and_wait(
+            1,
+            list(tm_allowed_actions),
+            tag="from_pbench_client",
+        )
+        self.sig_pub = sig_pub
+        self.logger.debug("constructed SignalExporter() object")
         return self
 
     def __exit__(self, *args):
-        """On context exit, just close down the to client channel object."""
-        if self.to_client_chan is not None:
-            self.to_client_chan.close()
+        """On context exit, just close down the to SignalExporter object."""
+        self.sig_pub.shutdown()
 
-    def publish(self, group, directory, action, args=None):
-        """publish - marshal a JSON document formed from the group, directory,
-        action, and args arguments.
+    def publish(
+        self,
+        group: str,
+        directory: Union[str, Path, None],
```

Comment:
That's OK, although I would have said `Optional[Union[str, Path]]` which I think is a bit more clear.